### ISSUE TITLE
Exponential Moving Average implementation for Vector Quantizer

### ISSUE THREAD

**Issue by willxxy** [evasnow1992/S-VQ-VAE#3]
Hello,

Thanks for the great work again. In a widely used [vqvae implementation](https://github.com/zalandoresearch/pytorch-vq-vae/blob/master/vq-vae.ipynb), they implemented the exponential moving average (EMA) for faster and more stable training.

To extend their idea to here, I have implemented the EMA in the `VectorQuantizer` class:

```
class VectorQuantizer(nn.Module):
    def __init__(self, num_embeddings, embedding_dim, commitment_cost, divergence_cost):
        super(VectorQuantizer, self).__init__()
        
        self._embedding_dim = embedding_dim
        self._num_embeddings = num_embeddings
        
        self._embedding = nn.Embedding(self._num_embeddings, self._embedding_dim)
        self._embedding.weight.data.uniform_(-1/self._num_embeddings, 1/self._num_embeddings)
        self._commitment_cost = commitment_cost
        self._divergence_cost = divergence_cost
        
        ### ADDED
        self.register_buffer('_ema_cluster_size', torch.zeros(num_embeddings))
        self._ema_w = nn.Parameter(torch.Tensor(num_embeddings, self._embedding_dim))
        self._ema_w.data.normal_()
        self._epsilon = 1e-5

        self._decay = 0.99
        ###

    def forward(self, inputs, label):
        # Calculate distances
        distances = (torch.sum(inputs**2, dim=1, keepdim=True) 
                    + torch.sum(self._embedding.weight**2, dim=1)
                    - 2 * torch.matmul(inputs, self._embedding.weight.t()))
            
        # Encoding
        encoding_indices = torch.reshape(label,(label.shape[0], 1))
        encodings = torch.zeros(encoding_indices.shape[0], self._num_embeddings).to(device)
        encodings.scatter_(1, encoding_indices, 1)
        
        close_indices = torch.argmin(distances, dim=1).unsqueeze(1)
        close_encodings = torch.zeros(close_indices.shape[0], self._num_embeddings).to(device)
        close_encodings.scatter_(1, close_indices, 1)
        
        indicator = 1 - (encoding_indices == close_indices).int()
        indicator = indicator.float()
        
        # Quantize
        quantized = torch.matmul(encodings, self._embedding.weight)
        close_quantized = torch.matmul(close_encodings, self._embedding.weight)
        
        ### ADDED
        if self.training: # defined by model.train() = True and model.eval() = False
            self._ema_cluster_size = self._ema_cluster_size * self._decay + \
                                     (1 - self._decay) * torch.sum(encodings, 0)
            
            # Laplace smoothing of the cluster size
            n = torch.sum(self._ema_cluster_size.data)
            self._ema_cluster_size = (
                (self._ema_cluster_size + self._epsilon)
                / (n + self._num_embeddings * self._epsilon) * n)
            
            dw = torch.matmul(encodings.t(), inputs)
            self._ema_w = nn.Parameter(self._ema_w * self._decay + (1 - self._decay) * dw)
            
            self._embedding.weight = nn.Parameter(self._ema_w / self._ema_cluster_size.unsqueeze(1))
        ###
        
        # Loss
        q_latent_loss = torch.mean((quantized - inputs.detach())**2)
        e_latent_loss = torch.mean((quantized.detach() - inputs)**2)
        x_latent_loss = torch.mean(indicator * ((close_quantized - inputs.detach())**2))
        d_latent_loss = torch.mean(indicator * ((close_quantized.detach() - inputs)**2))
        loss = q_latent_loss + self._commitment_cost * e_latent_loss - x_latent_loss - self._divergence_cost * d_latent_loss
        
        quantized = inputs + (quantized - inputs).detach()
        avg_probs = torch.mean(encodings, dim=0)
        perplexity = torch.exp(-torch.sum(avg_probs * torch.log(avg_probs + 1e-10)))
        
        return loss, quantized, perplexity, encodings
```
I trained on the MNIST example and got the following curves:

![output](https://github.com/evasnow1992/S-VQ-VAE/assets/90741489/c061c237-072d-4552-9d2d-787e17fa9db3)

I also got a test loss of Test loss: 0.2189 and test dataset acc of 0.848. 
Additionally, the time spent training with and without ema had a difference of about 2.1 seconds for 24 epochs.
